lib/node_modules/@stdlib/fs/resolve-parent-paths-by/lib/main.js

```diff
@@ -0,0 +1,507 @@
+	* @returns {void}
+	*/
+	function onExists( error, bool ) { // eslint-disable-line node/handle-callback-err
+		if ( bool ) {
+			return test( spath, onTest );
+		}
+		next();
+	}
+
+	/**
+	* Resolves the next candidate path.
+	*
+	* @private
+	* @returns {void}
+	*/
+	function next() {
+		if ( idx === paths.length ) {
+			// Resolve a parent directory:
+			child = dir;
+			dir = resolve( dir, '..' );
+
+			if ( child === dir ) {
+				return done( null, out );
+			}
+			idx = 0;
+		}
+		spath = resolve( dir, paths[ idx ] );
+		idx += 1;
+		exists( spath, onExists );
+	}
+
+	/**
+	* Callback invoked after testing a resolved path.
+	*
+	* @private
+	* @param {(Error|null)} error - error object
+	* @param {boolean} bool - boolean indicating if a path exists
+	* @returns {void}
+	*/
+	function onTest( error, bool ) {
+		if ( error ) {
+			return done( error );
+		}
+		if ( bool ) {
+			out.push( spath );
+			return done( null, out );
+		}
+		next();
+	}
+}
+
+/**
+* Asynchronously resolves one or more paths from a set of paths at a directory level according to a predicate function by walking parent directories.
+*
+* @private
+* @param {Array<string>} paths - paths to resolve
+* @param {string} dir - base directory
+* @param {Function} test - callback which tests whether a resolved path passes a test
+* @param {Callback} done - callback to invoke after resolving paths
+* @returns {void}
+*/
+function some( paths, dir, test, done ) {
+	var child;
+	var spath;
+	var FLG;
+	var out;
+
+	FLG = 0; // initialize flag to track if we are done traversing a directory level
+	out = [];
+
+	// Start at a base directory and continue moving up through each parent directory...
+	return next( dir );
+
+	/**
+	* Resolves paths within a directory.
+	*
+	* @private
+	* @param {string} dir - directory to search
+	*/
+	function next( dir ) {
+		var i;
+		for ( i = 0; i < paths.length; i++ ) {
+			spath = resolve( dir, paths[ i ] );
+			exists( spath, getCallback( spath ) );
+		}
+	}
+
+	/**
+	 * Returns a callback to be invoked upon checking for path existence.
+	*
+	* @private
+	* @param {string} spath - resolved path
+	* @returns {Callback} callback
+	*/
+	function getCallback( spath ) {
+		return onExists;
+
+		/**
+		* Callback invoked after testing a resolved path.
+		*
+		* @private
+		* @param {(Error|null)} error - error object
+		* @param {boolean} bool - boolean indicating if a path exists
+		* @returns {void}
+		*/
+		function onTest( error, bool ) {
+			if ( error ) {
+				return done( error );
+			}
+			if ( bool ) {
+				out.push( spath );
+			}
+		}
+
+		/**
+		* Callback invoked after checking for path existence.
+		*
+		* @private
+		* @param {(Error|null)} error - error object
+		* @param {boolean} bool - boolean indicating if a path exists
+		* @returns {void}
+		*/
+		function onExists( error, bool ) { // eslint-disable-line node/handle-callback-err
+			if ( bool ) {
+				test( spath, onTest );
+			}
+			FLG += 1;
```

Comment:
This logic seems off. `test` may be an asynchronous function. In which case, the `onTest` callback will be called after hitting `FLG += 1`. If done multiple times, that could mean that L203-221 could be exercised **before** any `onTest` call is complete. This race condition needs to be addressed.

Reply:
![image](https://github.com/user-attachments/assets/246fe7c9-9577-4653-8c94-057762684bdb)
@kgryte,  I wanted to clarify whether using async/await to wait for the test function to complete would be an appropriate solution here.

Reply:
No, no using async/await. This implementation should be entirely callback-based

Reply:
```javascript
function some( paths, dir, test, done ) {
	var child;
	var spath;
	var FLG;
	var out;

	FLG = 0; // initialize flag to track if we are done traversing a directory level
	out = [];

	// Start at a base directory and continue moving up through each parent directory...
	return next( dir );

	/**
	* Resolves paths within a directory.
	*
	* @private
	* @param {string} dir - directory to search
	*/
	function next( dir ) {
		var i;
		for ( i = 0; i < paths.length; i++ ) {
			spath = resolve( dir, paths[ i ] );
			exists( spath, getCallback( spath ) );
		}
	}

	/**
	 * Returns a callback to be invoked upon checking for path existence.
	*
	* @private
	* @param {string} spath - resolved path
	* @returns {Callback} callback
	*/
	function getCallback( spath ) {
		return onExists;

		/**
		* Callback invoked after testing a resolved path.
		*
		* @private
		* @param {(Error|null)} error - error object
		* @param {boolean} bool - boolean indicating if a path exists
		* @returns {void}
		*/
		function onTest( error, bool ) {
			if ( error ) {
				return done( error );
			}
			if ( bool ) {
				out.push( spath );
			}
			// Increment flag after `test` completes:
			FLG += 1;
			if ( FLG === paths.length ) {
				// Check if we have resolved any paths...
				if ( out.length > 0 ) {
					return done( null, out );
				}
				// Resolve a parent directory:
				child = dir;
				dir = resolve( dir, '..' );

				// Reset flag:
				FLG = 0;

				// If we have already reached root, we cannot resolve any higher directories...
				if ( child === dir ) {
					return done( null, out );
				}
				// Resolve paths at next directory level:
				return next( dir );
			}
		}

		/**
		* Callback invoked after checking for path existence.
		*
		* @private
		* @param {(Error|null)} error - error object
		* @param {boolean} bool - boolean indicating if a path exists
		* @returns {void}
		*/
		function onExists( error, bool ) { // eslint-disable-line node/handle-callback-err
			if ( bool ) {
				// Pass the path to the `test` function and use `onTest` as the callback:
				test( spath, onTest );
			} else {
				// Increment flag if the path does not exist:
				FLG += 1;
				if ( FLG === paths.length ) {
					// Check if we have resolved any paths...
					if ( out.length > 0 ) {
						return done( null, out );
					}
					// Resolve a parent directory:
					child = dir;
					dir = resolve( dir, '..' );

					// Reset flag:
					FLG = 0;

					// If we have already reached root, we cannot resolve any higher directories...
					if ( child === dir ) {
						return done( null, out );
					}
					// Resolve paths at next directory level:
					return next( dir );
				}
			}
		}
	}
}
```
Does this work out

Reply:
@kgryte, Am I doing it right in above mentioned programme, It passes all test cases & would pass the race conditions during async processing

Reply:
@gururaj1512 The updated logic seems apt to me. Maybe you can also refactor the code where you are checking whether to move to the next directory or stop (inside the `if ( FLAG === paths.length )` check), into a function to avoid code duplication, WDYT?. Logic wise though, seems reasonable to me.👍